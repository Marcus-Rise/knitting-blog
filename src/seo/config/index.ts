export * from "./seo-config.service.interface";
export * from "./seo-config.service.provider";
export * from "./seo-config.service";
