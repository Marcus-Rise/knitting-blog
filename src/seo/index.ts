export { SEO_SERVICE_PROVIDER } from "./seo.service.interface";
export type { ISeoService } from "./seo.service.interface";
export { SeoModule } from "./seo.module";
export type { ISeoConfigService } from "./seo-config.service.interface";
export { SEO_CONFIG_SERVICE_PROVIDER } from "./seo-config.service.interface";
