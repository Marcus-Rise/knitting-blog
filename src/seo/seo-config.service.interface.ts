const SEO_CONFIG_SERVICE_PROVIDER = Symbol("ISeoConfigService");

interface ISeoConfigService {
  readonly allowRobots: boolean;
}

export type { ISeoConfigService };
export { SEO_CONFIG_SERVICE_PROVIDER };
