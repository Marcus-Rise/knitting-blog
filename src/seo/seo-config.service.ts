import type { ISeoConfigService } from "./seo-config.service.interface";
import { injectable } from "inversify";

@injectable()
class SeoConfigService implements ISeoConfigService {
  readonly allowRobots: boolean = Boolean(process.env.ALLOW_ROBOTS);
  readonly googleVerificationCode: string = process.env.NEXT_PUBLIC_GOOGLE_VERIFICATION ?? "";
  readonly yandexVerificationCode: string = process.env.NEXT_PUBLIC_YANDEX_VERIFICATION ?? "";
}

export { SeoConfigService };
