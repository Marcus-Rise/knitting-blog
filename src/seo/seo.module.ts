import { ContainerModule } from "inversify";
import type { ISeoService } from "./seo.service.interface";
import { SEO_SERVICE_PROVIDER } from "./seo.service.interface";
import { SeoServiceMock } from "./seo.service.mock";

const SeoModule = new ContainerModule((bind) => {
  bind<ISeoService>(SEO_SERVICE_PROVIDER).to(SeoServiceMock).inSingletonScope();
});

export { SeoModule };
