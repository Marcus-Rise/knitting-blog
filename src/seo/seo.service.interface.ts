const SEO_SERVICE_PROVIDER = Symbol("ISeoService");

interface ISeoService {
  generateSitemap(hostName: string): Promise<string>;
  generateRobotsTxt(): Promise<string>;
}

export type { ISeoService };
export { SEO_SERVICE_PROVIDER };
