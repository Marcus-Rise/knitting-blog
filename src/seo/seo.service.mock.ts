import type { ISeoService } from "./seo.service.interface";
import { injectable } from "inversify";

@injectable()
class SeoServiceMock implements ISeoService {
  generateRobotsTxt(): Promise<string> {
    return Promise.resolve("");
  }

  generateSitemap(): Promise<string> {
    return Promise.resolve("");
  }
}

export { SeoServiceMock };
