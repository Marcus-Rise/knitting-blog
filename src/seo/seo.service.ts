import type { ISeoService } from "./seo.service.interface";
import { inject, injectable } from "inversify";
import type { IPostService } from "../post";
import { POST_SERVICE_PROVIDER } from "../post";

@injectable()
class SeoService implements ISeoService {
  constructor(@inject(POST_SERVICE_PROVIDER) private readonly posts: IPostService) {}

  async generateRobotsTxt(hostName: string): Promise<string> {
    const sections: { [key: string]: string } = {
      "User-agent": "*",
      Disallow: "/",
      Sitemap: `https://${hostName}/sitemap.xml`,
    };

    return Object.keys(sections)
      .map((i) => `${i}: ${sections[i]}`)
      .join("\n");
  }

  async generateSitemap(hostName: string): Promise<string> {
    let buf = `<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">`;
    const posts = await this.posts.getList(0);

    posts.forEach((post) => {
      const url = `https://${hostName}/${post.slug}`;
      const period = "daily";
      const priority = 0.9;

      buf += `<url><loc>${url}</loc><changefreq>${period}</changefreq><priority>${priority}</priority></url>`;
    });

    buf += `</urlset>`;

    return buf.toString();
  }
}

export { SeoService };
