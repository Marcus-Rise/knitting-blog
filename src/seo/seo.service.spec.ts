import "reflect-metadata";
import { SeoService } from "./seo.service";
import { mock } from "jest-mock-extended";
import type { IPost, IPostService } from "../post";

describe("SeoService", () => {
  test("generateRobotsTxt", async () => {
    const service = new SeoService(mock<IPostService>());

    const data = await service.generateRobotsTxt("hostname");

    expect(data.length).toBeGreaterThan(0);
    expect(data).toMatchSnapshot();
  });

  test("generateSitemap", async () => {
    const service = new SeoService(
      mock<IPostService>({
        getList: () =>
          Promise.resolve([
            mock<IPost>({
              slug: "post1",
            }),
            mock<IPost>({
              slug: "post2",
            }),
          ]),
      }),
    );

    const data = await service.generateSitemap("hostname");

    expect(data.length).toBeGreaterThan(0);
    expect(data).toMatchSnapshot();
  });
});
