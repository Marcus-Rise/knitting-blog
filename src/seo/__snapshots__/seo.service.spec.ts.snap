// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeoService generateRobotsTxt 1`] = `
"User-agent: *
Disallow: /
Sitemap: https://hostname/sitemap.xml"
`;

exports[`SeoService generateSitemap 1`] = `"<urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\"><url><loc>https://hostname/post1</loc><changefreq>daily</changefreq><priority>0.9</priority></url><url><loc>https://hostname/post2</loc><changefreq>daily</changefreq><priority>0.9</priority></url></urlset>"`;
